[lldb] Adding A new Binding helper for JSONTransport.

ashgti opened the pull request:
This adds a new Binding helper class to allow mapping of incoming and outgoing requests / events to specific handlers.

This should make it easier to create new protocol implementations and allow us to create a relay in the lldb-mcp binary.

ashgti committed:
[lldb] Adding A new Binding helper for JSONTransport.

This adds a new Binding helper class to allow mapping of incoming and outgoing requests / events to specific handlers.

This should make it easier to create new protocol implementations and allow us to create a relay in the lldb-mcp binary.

ashgti committed:
Adding some comments and adjusting the naming of the helpers.

JDevlieghere reviewed (commented):
I did a first pass but I definitely need to do another pass with everything I learned during the first round, but don't want to hold back initial comment on that.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -100,22 +175,301 @@ template <typename Req, typename Resp, typename Evt> class Transport {
+  class Binder;
+  using BinderUP = std::unique_ptr<Binder>;
+
+  /// Binder collects a table of functions that handle calls.
```
> By defining message types we can get that automatically out of the messages, and then have it call the right callback?

Yea, for example, from the Server.cpp
```
MCPTransport::BinderUP Server::Bind(MCPTransport &transport) {
  MCPTransport::BinderUP binder =
      std::make_unique<MCPTransport::Binder>(transport);
  binder->bind<InitializeResult, InitializeParams>(
      "initialize", &Server::InitializeHandler, this);
  binder->bind<ListToolsResult, void>("tools/list", &Server::ToolsListHandler,
                                      this);
  binder->bind<CallToolResult, CallToolParams>("tools/call",
                                               &Server::ToolsCallHandler, this);
  binder->bind<ListResourcesResult, void>("resources/list",
                                          &Server::ResourcesListHandler, this);
  binder->bind<ReadResourceResult, ReadResourceParams>(
      "resources/read", &Server::ResourcesReadHandler, this);
  binder->bind<void>("notifications/initialized",
                     [this]() { Log("MCP initialization complete"); });
  return binder;
}
```

The `bind` method setups the serialization of the Params and Results and wraps them into the `Request` and `Response` types.

This means the server doesn't need to implement its own method dispatching mechanism, that can be handled by the `Binder` and Server can be simplified a bit. You can also see an example in JSONTransportTest.cpp that does this simple binding of a incoming call, outgoing call, incoming event and outgoing event.

> does this need to be tied to JSONTransport? Can this be built on top of an arbitrary transport?

Within each `bind` call we do assume that `Params` and `Result` types are all JSON serializable.

That could be pulled apart a bit more though if we wanted to remove the JSON requirement. That could become a parameter to the template I suppose.

ashgti replied:
Also, another benefit of the binder not being part of the Server implementation here specifically is that the Server can choose if it wants to handle a single client or multiple clients per instance. We can bind the methods with additional context parameters to make the server independent of the transport.

ashgti committed:
Trying to improve the organization of the JSONTransport file.

Moved the Binder below the rest of the transport classes and moved some of the implementations below the class definition to improve readability of the Binder interface without seeing a lot of template details.

ashgti committed:
Fixing clang-format.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -50,17 +56,88 @@ class TransportUnhandledContentsError
+  std::error_code convertToErrorCode() const override;
+
+private:
+  std::string m_method;
```
I added a comment about this, this comes from the Binder. In JSONRPC the refer to the remote calls as 'methods'. In the DAP, which isn't precisely a JSONRPC transport, its a 'command' or an 'event', so the terminology isn't consistent. I went with the JSONRPC based name.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -50,17 +56,88 @@ class TransportUnhandledContentsError
   std::string m_unhandled_contents;
 };
 
+class InvalidParams : public llvm::ErrorInfo<InvalidParams> {
```
Added comments to this and a bunch of other classes in this file.

JDevlieghere reviewed (commented):
Thanks, this is looking good. Left some nits, but I think I have a good grasp on the Binder now. I assume the plan is to use the Binder for `lldb-dap` too?

ashgti committed:
Consolidating protocol descriptions into a single 'concept'.

The JSONTransport having to carry around 4 template parameters was a lot of repeated information.

To try to simplify this, I created a 'concept' for `ProtocolDescriptor` that describes the single type definition used to describes a protocol.
The `BindingBuilder` concept is an extension on `ProtocolDescriptor` that has additional helpers for using the Binder.

This helps relate the concepts together into a single place, which I find helpful. Previously, there were a few free standing template functions that needed to be implemented for the Binder to work.

JDevlieghere reviewed (commented):
Still some outstanding comments, but once these addressed I thin this is good to go.

ashgti committed:
Improving comment formatting.

ashgti committed:
Wrapping concepts in an #ifdef instead of commenting them out.

ashgti committed:
Fixing a c++17 compatibility.

ashgti committed:
Merge branch 'main' into lldb-jsontransport-binder

ashgti commented:
Taking a look.

DavidSpickett commented:
> It is an assert inside CRT. Probably states of handles are wrong.

Or that's just the assert going off and using something in CRT to do so.

```
Expected: succeeded
  Actual: failed  (Unsupported file type 2)
```
This could be coming from `lldb/source/Host/windows/MainLoopWindows.cpp`:
```
    DWORD file_type = GetFileType(waitable_handle);
    if (file_type != FILE_TYPE_PIPE) {
      error = Status::FromErrorStringWithFormat("Unsupported file type %ld",
                                                file_type);
      return nullptr;
    }
```
https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfiletype

These are test cases, so it would make some sense if one of the streams was not a pipe but some premade string or text file.
